omniduct/filesystems/s3.py

```diff
@@ -189,11 +191,73 @@ def _dir(self, path):
     #         if pattern is None or pattern.match(k.key[len(path_prefix):]):
     #             yield k.key
 
+    def _format_s3_path(self, path):
+        return 's3:' + path[4:] if path.startswith('s3n:') or path.startswith('s3a:') else path
+
+    def _is_s3(self, path):
+        """This function gives True if the given path is valid s3 path. This is used for the purpose of
+        validation."""
+        return 's3' in self._format_s3_path(path)
+
+    def _bucket_name(self, path):
+        """This function gives the bucket name.
+        Example: If the path s3://bucket_name/folder1/folder2/ is given as a parameter to this function
+        it returns bucket_name"""
+        if self._is_s3(path):
+            path = urlparse(path)
+            return path.netloc
+        else:
+            raise FileNotFoundError("path `{}` does not exist.".format(path))
+
+    def _file_or_dirpath(self, path):
+        """This function gives the directory or the filename.
+        Example: If the path s3://bucket_name/folder1/folder2/file1.json is given as a parameter to this function
+        it returns folder1/folder2/file1.json"""
+        if self._is_s3(path):
+            path = urlparse(path)
+            return path.path
+        else:
+            raise FileNotFoundError("path `{}` does not exist.".format(path))
+
+    def _object_exists(self, path):
+        """This function checks if an object exists in s3"""
+        s3 = self._resource
+        try:
+            s3.Object(self._bucket_name(path), self._file_or_dirpath(path)).load()
+            return True
+        except:
+            return False
+
     def _mkdir(self, path, recursive):
-        raise NotImplementedError("Support for S3 write operations has yet to be implemented.")
+        assert self._is_s3(path), 'path must be valid s3 path'
+        bucket = self._bucket_name(path)
+        file_or_dirname = self._file_or_dirpath(path)
+        if not self._object_exists(path):
+            self._client.put_object(Bucket=bucket, Key=file_or_dirname)
+
+    def _remove_file(self, path):
+        """Removes a single key or file from s3"""
+        assert self._is_s3(path)
+        s3 = self._resource
+        bucket_name = self._bucket_name(path)
+        file_or_dir_name = self._file_or_dirpath(path)
+        s3.Object(bucket_name, file_or_dir_name).delete()
+        return True
 
     def _remove(self, path, recursive):
-        raise NotImplementedError("Support for S3 deletion operations has yet to be implemented.")
+        """This function deletes a single file or object from s3 if the recursive flag is False else it
+        deletes the objects or files under the given directory recursively."""
```

Comment:
Similar notes about reusing existing methods here.

Reply:
can I delete this description?

Reply:
deleted the description and unwanted portion